src/use-swr.ts

```diff
@@ -571,49 +571,37 @@ function useSWR<Data = any, Error = any>(
     addRevalidator(RECONNECT_REVALIDATORS, onReconnect)
     addRevalidator(CACHE_REVALIDATORS, onUpdate)
 
-    return () => {
-      // cleanup
-      dispatch = () => null
-
-      // mark it as unmounted
-      unmountedRef.current = true
-
-      removeRevalidator(FOCUS_REVALIDATORS, onFocus)
-      removeRevalidator(RECONNECT_REVALIDATORS, onReconnect)
-      removeRevalidator(CACHE_REVALIDATORS, onUpdate)
-    }
-  }, [key, revalidate])
-
-  // set up polling
-  useIsomorphicLayoutEffect(() => {
-    let timer = null
     const tick = async () => {
       if (
         !stateRef.current.error &&
-        (config.refreshWhenHidden || isDocumentVisible()) &&
-        (config.refreshWhenOffline || isOnline())
+        (configRef.current.refreshWhenHidden || isDocumentVisible()) &&
+        (configRef.current.refreshWhenOffline || isOnline())
       ) {
         // only revalidate when the page is visible
         // if API request errored, we stop polling in this round
         // and let the error retry function handle it
         await revalidate({ dedupe: true })
       }
-      if (config.refreshInterval) {
-        timer = setTimeout(tick, config.refreshInterval)
+      if (configRef.current.refreshInterval) {
+        setTimeout(tick, configRef.current.refreshInterval)
       }
     }
-    if (config.refreshInterval) {
-      timer = setTimeout(tick, config.refreshInterval)
+    if (configRef.current.refreshInterval) {
+      setTimeout(tick, configRef.current.refreshInterval)
     }
+
     return () => {
-      if (timer) clearTimeout(timer)
+      // cleanup
+      dispatch = () => null
+
+      // mark it as unmounted
+      unmountedRef.current = true
+
+      removeRevalidator(FOCUS_REVALIDATORS, onFocus)
+      removeRevalidator(RECONNECT_REVALIDATORS, onReconnect)
+      removeRevalidator(CACHE_REVALIDATORS, onUpdate)
```

Comment:
I think we should set up a `timer = setTimeout(...)` and clear it when exiting the effect here.